o/i/a/common: allow groups in path patterns and adjust pattern precedence

olivercalder opened the pull request:
Allow arbitrary groups, wildcards, doublestars, and `'?'` characters in path patterns. Groups may be nested arbitrarily as well.

The only remaining restriction on the full set of path patterns supported by doublestar (https://pkg.go.dev/github.com/bmatcuk/doublestar/v4#Match) is that character classes (enclosed by `'['` and `']'`) are not supported. Defining precedence and pattern conflicts and allowing for excluded character classes is beyond the scope of apparmor prompting, at the moment.

olivercalder committed:
o/i/a/common: allow groups in path patterns

Path patterns can be thought of as a base path concatenated with either
a suffix or a group over multiple suffixes.

The list of suffixes (by precedence) is:

1. ``           (no suffix has highest priority)
2. `/*.ext`     (particular file extension in base path directory)
3. `/**/*.ext`  (particular file extension in any subdirectory of base path)
4. `/.*`        (dotfiles in base path directory)
5. `/**/*.`     (dotfiles in any subdirectory of base path)
6. `/*`         (any file in base path directory)
7. `/**`        (any file in any subdirectory of base path)

When a path is received, it is matched against existing rules, each of
which has a path pattern. If there are any groups in a rule's path
pattern, that pattern has been expanded (at time of rule being added)
into a pattern without groups for each option in the group of the
original path pattern.

When the path incoming path is matched against existing rules, it may
match more than one path pattern. If there is only one pattern which is
of the form with the highest precedence (of the set of matching
patterns), then that rule has precedence. Otherwise, the longest pattern
has precedence, as all suffixes of a given precedence matching a given
rule must be identical, so the base path of the longest pattern must be
longest, and thus the most specific.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: allow multi-part file extensions in path patterns

In order to support multi-part file extensions, such as `.tar.gz`, in
path patterns, we need to change the way precedence is computed for
patterns in the same suffix-precedence category.

For example, the following have the same suffix type (`/**/*.ext`), but
have equal length, so the old way of computing precedence by pattern
length fails:

* `/foo/bar/baz/**/*.gz`
* `/foo/bar/**/*.tar.gz`

In the above, we see that the first rule has a more specific base path,
while the second rule has a more specific file extension. It is a bit
arbitrary whether base path or file extension is more important, but for
consistency's sake, we introduce a new principle to path matching:

**Within a given suffix-precedence category, base path specificity is
more important than suffix specificity.**

Now, (1) has higher precedence than (2), despite that the latter has a
longer total pattern:

1. `/foo/bar/x/**/*.gz`
2. `/foo/bar/**/*.tar.gz

The precedence relationship still holds between suffix categories:

1. `/foo/bar/*.gz`
2. `/foo/bar/**/*.tar.gz`
3. `/foo/bar/*`
4. `/foo/bar/**`

As before, (1) has precedence over (2), because it matches files with a
given extension in the base path directory, rather than files with a
(more specific) file extension in any subdirectory, since any
subdirectory is a lower-precedence suffix category than in-directory.
The rules 1-4 are ordered in terms of precedence.

Note: the following should never be compared, since they cannot match
the same path, but if they *were* to be compared, the precedence order
would be as follows:

1. `/foo`
2. `/foo/bar/*.gz`
3. `/foo/*.tar.gz`
4. `/foo/bar/baz/**/*.gz`
5. `/foo/bar/**/*.tar.gz`
6. `/foo/bar/baz/fizz/*`
7. `/foo/bar/baz/*`
8. `/foo/bar/baz/fizz/buzz/**`
9. `/foo/bar/baz/fizz/**`

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: adjust precedence and allow greater wildcard flexibility

Precedence has been adjusted so that *where* has priority over *what*.
That is, base path length takes priority over the specificity of file
extensions or other suffixes. If two patterns have the same base path,
but one follows with `/*` while the another follows with `/**`, the
former has precedence, but if a base+`/*` rule has a shorter base path
than a base+`/**` rule, the latter has precedence.

For example, the pattern `/foo/bar/**` has precedence over /foo/bar*`,
since the former requires that the path ends in `/` (TODO: this isn't
actually the case yet, since path matching hasn't yet been updated to be
more strict now that paths and path patterns may end in `/` and groups
are allowed in path patterns). Both of these patterns match `/foo/bar/`
(and also, currently, incorrectly, `/foo/bar` -- this will be fixed).

Wildcards are no longer restricted to being at the end of the path
pattern, grouped together and followed by a file extension or nothing at
all. Now, characters and/or directories can appear arbitrarily around
the single `*`, and after the `/**/`. However, there must still be at
most one `/**/` and one `*` in the pattern, with the `*` occurring after
the `/**/` if it exists.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: allow arbitrary wildcards in path patterns

There are no longer restrictions about where wildcards or doublestars
may occur in path patterns. Prior modifications to precedence
computation which placed priority on *where* instead of *what* mean that
precedence between any two patterns (without groups or classes) which
match the same path can be decided.

More broadly, the concept of suffixes is no longer used in precedence
computation (due to the previous change) or checking the validity of
path patterns (new in this commit).

As such, the allowable path patterns glob can be modified to be far more
flexible, with wildcards occurring anywhere and the group beginning
anywhere. Additionally, we now allow "mis-use" of `**`, such as
concatenated to a non-`/` character, which is equivalent to `*`. Since
the group may now begin anywhere, we also allow repeated `//`, which is
equivalent to `/`, since it is difficult to check for these prior to
expansion. In both cases, `ExpandPathPattern()` replaces these repeated
characters with a single `*` or `/`, respectively, so that conflicts
with other equivalent patterns still occur as usual.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: fixed path pattern precedence

Re-architected the pattern precedence system, increasing performance,
clarity, and correctness. Notably, adjusted the precedence of patterns
with a glob as the next character compared to other patterns with a path
separator as the next character.

The new precedence order is as follows (highest to lowest precedence):

literals:
- /foo/bar/baz
- /foo/bar/
terminated:
- /foo/bar
singlestars:
- /foo/bar/*/baz
- /foo/bar/*/
- /foo/bar/*/*baz
- /foo/bar/*/*
- /foo/bar/*
- /foo/bar/*/**
glob:
- /foo/bar*baz
- /foo/bar*/baz
- /foo/bar*/baz/**
- /foo/bar*/
- /foo/bar*/*baz
- /foo/bar*/*/baz
- /foo/bar*/*/
- /foo/bar*/*
- /foo/bar*
doublestars:
- /foo/bar/**/baz
- /foo/bar/**/*baz/
- /foo/bar/**/*baz
terminal doublestar:
- /foo/bar/**/        # These are tough... usually, /foo/bar/**/ would have precedence over
- /foo/bar/**/*       # precedence over /foo/bar/**/*baz, but in this case,
- /foo/bar/**         # the trailing *baz adds more specificity.
glob with immediate doublestar:
- /foo/bar*/**/baz
- /foo/bar*/**/
- /foo/bar*/**

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: refactored and simplified pattern precedence

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder commented:
If you'd like to take a look @TheSignPainter98, I finally got path pattern precedence working, and without regular expressions!

olivercalder committed:
o/i/a/common: replaced usage of `strings.CutSuffix`

The `strings.CutSuffix` function requires go 1.20, which is not yet
required for snapd.

Furthermore, using the `CutSuffix` function in this case means calling
`HasSuffix` with `"/**/*"`, then constructing a string with the
`"/**/*"` removed, then constructing a new string with the `"/**"`
re-appended. Not very efficient.

Instead, simply check if the pattern has the `"/**/*"` suffix, and if
so, strip only the final `"/*"`.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: allow arbitrary nested groups in path patterns

The `ExpandPathPattern` function has been rewritten to allow arbitrary
usage of groups in path patterns, without restrictions on the number of
groups which may occur nor the depth to which groups may be nested.

Now, the only restrictions placed on path patterns are that they must
begin with a `'/'` character and must not use unescaped `'['`, `']'`, or
`'?'` characters, as the current way precedence is computed relies on
being able to fully expand all path patterns into literals and
wildcards, and expanding `'[a-z]`' classes is not yet supported. If
classes were not expanded, they could be treated similarly to a single
`'?'` character with more restrictions, but computing precedence between
two classes is difficult, if not impossible.

As before, a `'\'` in front of any special character causes it to be
treated as a literal.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: allow `'?'` characters in path patterns

Any single character (denoted `'?'`) should have lower precedence than a
literal character, but should have higher precedence than zero or more
characters (denoted `'*'`).

There are some tricky interactions between `'?'` and `'*'`, so added
some unit tests to make sure these are cleaned properly. Additionally,
improved the robustness of path pattern cleaning in general.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	return nil, 0, fmt.Errorf(`invalid path pattern: unmatched '{' character: %q`, pattern)
+}
+
+var (
+	duplicateSlashes    = regexp.MustCompile(`(^|[^\\])/+`)
+	charsDoublestar     = regexp.MustCompile(`([^/\\])\*\*+`)
+	doublestarChars     = regexp.MustCompile(`([^\\])\*\*+([^/])`)
+	duplicateDoublestar = regexp.MustCompile(`/\*\*(/\*\*)+`) // relies on charsDoublestar running first
+	starsAnyMaybeStars  = regexp.MustCompile(`([^\\])\*+(\?\**)+`)
+)
+
+func cleanPattern(pattern string) string {
+	pattern = duplicateSlashes.ReplaceAllString(pattern, `${1}/`)
+	pattern = charsDoublestar.ReplaceAllString(pattern, `${1}*`)
+	pattern = doublestarChars.ReplaceAllString(pattern, `${1}*${2}`)
+	pattern = duplicateDoublestar.ReplaceAllString(pattern, `/**`)
+	pattern = starsAnyMaybeStars.ReplaceAllStringFunc(pattern, shiftStarAfterAnyChars)
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-2]
+	}
+	return pattern
+}
+
+func shiftStarAfterAnyChars(s string) string {
+	return strings.Map(deleteStars, s) + "*"
+}
+
+func deleteStars(r rune) rune {
+	if r == '*' {
+		return -1
+	}
+	return r
+}
```
This is rather inefficient. I think @TheSignPainter98 mentioned some way to do a series of regex matches/substitutions on a single pass over a string. Any ideas how this could be improved?

TheSignPainter98 replied:
This may be possible with a regex set (something like Rust's [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html)), but you'd need knowledge of which regex matched, where and what was the length of the match. I can't seem to see these in Go, but iirc as this is a one-time cost, perhaps multiple passes with a few regexes isn't too bad? Personally I'd prefer a loop though

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
```
Hold on, `defaultPrefixes` is `[]string{""}`, is that valid? I may be missing some context, but it looks a little odd to me. Every time you use `defaultPrefixes`, could you use `[]string{}` or `[]string(nil)` instead? I think this would reduce the extra overhead of needing to understand this (admittedly quite simple) default value :P

This change would also make the zero-expansion case simpler.

Also, to more clearly signal the intent, could 'expanded' be called something like `ret` or `res` to show more clearly that this is what's being built up for the return?

olivercalder replied:
Yes, it must be a list with one item, as this is the list of previous patterns to which anything new should be appended. If it has length 0, there are no previous patterns, and thus the function would always return zero patterns. It's so that we can use the `expanded` variable throughout the function and continually add the previously-expanded components to it.

Essentially, whenever we encounter a group, we expand it, and then for each newly expanded path in that group, and for each previously expanded path in `expanded`, we adjoin the previous one to the new one. Then, at the end, we return `expanded`.

Do you think it would be clearer to rename `expanded` to `previouslyExpanded` or `expandedPrefixes`?

olivercalder replied:
I've renamed this to `currPrefixes`, and its behavior now matches that in `expandPathPatternFromIndex`.

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	currLiteralStart := 0
+	for i := 0; i < patternLen; i++ {
+		if pattern[i] == '\\' {
+			i += 1
```
Just a nitpick, I'd put the state transition `i += 1`  at the end of this `if` (as in so the early return comes earlier)

olivercalder replied:
The early return returns if the incremented `i` pushes the index beyond the end of the pattern, so I need to increment it first.

Alternatively, this would work:
```go
if i+1 == patternLen {
    return nil, fmt.Errorf(...)
}
i += 1
continue
```

but this seems a bit less clean, perhaps? We also can't simply continue here and let the loop catch the index out of bounds, since then it will ignore the trailing `'\\'` instead of returning an error.

Do you prefer the latter suggested block?

olivercalder replied:
Moving the check for trailing `\` up fixes this.

TheSignPainter98 replied:
Fair enough, you could remove the `+1` from the check by using `>` :P

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -403,89 +722,33 @@ func ValidateConstraintsOutcomeLifespanDuration(iface string, constraints *Const
+// ensure that patterns without trailing slashes match paths with trailing
+// slashes. However, patterns with trailing slashes should not match paths
+// without trailing slashes.
+func PathPatternMatches(pattern string, path string) (bool, error) {
```
Not sure whether this is consistent with elsewhere, but I see `doublestar` calls these methods 'Match,' so perhaps this should be `PathPatternMatch`? Otherwise it looks like it should return a slice of something :P

olivercalder replied:
Good point, thanks!

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -403,89 +722,33 @@ func ValidateConstraintsOutcomeLifespanDuration(iface string, constraints *Const
+	return doublestar.Match(patternSlash, path)
 }
+
+var doublestarSuffix = regexp.MustCompile(`(/\*\*)?/?$`)
```
Just a nitpick, when first reading `PathPatternMatches`, I wasn't sure where `doublestarSuffix` came from, so I'd prefer if this were defined before all functions it's used in :P

olivercalder replied:
Fair enough. If it were a function, I'd want it where it is, but since it's a variable you're right that it's more confusing.

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+
+	// The following matches valid path patterns
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
```
I think the `problematicChars` set is a little smaller than it used to be, perhaps the inlining version is okay? I think the extra complexity of extra variables (which aren't used elsewhere) is a little much. Perhaps inlining were still too complicated, a `func init()` could scope things more nicely

olivercalder replied:
Good point, I agree that inlining makes the most sense here, given it's just two characters left (albeit escaped).

olivercalder replied:
Actually, it could all be inlined into:

```go
var allowablePathPatternRegexp = regexp.MustCompile(`^/([^\[\]]|\\[\[\]])*$`)
```

I think this is nicer than having a bunch of variables and string formatting.

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+		if pattern[i] != '{' {
+			continue
+		}
+		infix := pattern[currLiteralStart:i]
```
I'd probably move this definition a little later as it isn't needed just yet here :P

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
```
In general, it's okay not to worry about caching `patternLen` as this is just a single go-assembly instruction to read a field. This would also avoid the need to avoid passing it around next to `pattern` :P

olivercalder replied:
Good to know, that's much nicer, thanks!

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
+	currLiteralStart := 0
+	for i := 0; i < patternLen; i++ {
```
This will iterate through by bytes rather than runes, is this okay? Perhaps `for i, c := range pattern {` could work? Or at least  `for i := range pattern { c := pattern[i]`.

Either way, storing `pattern[i]` would clean things up a little

olivercalder replied:
I think I need to iterate by runes, in case path patterns are non-ASCII. But I need a way to skip the next rune if it's preceded by a `'\\'` -- that's why I didn't iterate over `i, c`, and why I didn't define `c := pattern[i]`, since it might immediately be reassigned and I only use it 2 times after that anyway. I suppose a variable would do, but I still need a way of creating a substring starting after a particular rune index and continuing until another index.

Wouldn't `for i, c := range pattern` be essentially identical to `for i := range pattern`, at least in terms of byte vs rune iteration?

TheSignPainter98 replied:
Ah my mistake, I thought one iterated bytes but they both iterate runes

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+// seen, returns an error.
+func expandPathPatternFromIndex(pattern string, index, patternLen int) (expanded []string, end int, err error) {
+	// Record total list of expanded patterns, to which other lists are appended
+	expanded = []string{}
```
Named returns are good for documentation, but it's usually better to shadow them---for a moment I thought there was a global called 'expanded'!

Also, the initial value of this `expanded` is slightly different from the one in the above function `[]string{""}`

olivercalder replied:
Yes, within a group vs between groups I'm handling things a bit differently. Every time a comma (or '}') is seen, all the expanded patterns for the current group option are appended to the previous ones, whereas between groups, we combine the previosly-expanded and newly-expanded patterns via a Cartesian product of sorts.

I agree, the names I used are inconsistent. I'm going to rename `expanded` in `ExpandPathPattern` to `currPrefixes`. But I'm not sure what you mean in this case... I'm returning a named value (or `nil`) for every return statement, but do you mean that expanded should be `var expanded []string`? This results in an error: `expanded redeclared in this block`.

TheSignPainter98 replied:
Oh no I mean `expanded := []string{}` should work unless I'm mistaken

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+//	- /foo/bar*/**/baz
+//	- /foo/bar*/**/
+//	- /foo/bar*/**
+func GetHighestPrecedencePattern(patterns []string) (string, error) {
```
Is it necessary to both sort and obtain the max priority pattern? If you only need one, would maintaining some `maxPriorityPattern` each iteration work?

olivercalder replied:
There could be multiple patterns with the same priority at a given iteration, so all top priority patterns need to be stored until the next iteration, when we look at the next character. I don't think sorting comes into play here.

Technically, the patterns could probably be sorted to get a good idea of which probably has the highest priority, but with escaped characters and the special cases around singlestars and doublestars, this would not really help and might distract from the checks which really matter.

I'm not fully sure I understand what you mean though

TheSignPainter98 replied:
Ah I may have misunderstood the structure which is being maintained, the naming made me think the entire set of patterns would be entered eventually, hence being some kind of sort. Perhaps a slice would be cleaner for this?

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	}
+	p := ""
+	for pattern := range remainingPatterns {
+		p = pattern
```
Wait is this just getting the last one? :P

olivercalder replied:
It's getting the only one. The loop only ends when `len(remainingPatterns) == 1` (it can't be zero). But I need a guaranteed return outside the loop, so I threw a dummy `p := ""` there and assigned it in the "loop" so that the return could happen after the loop.

TheSignPainter98 replied:
Er this is quite strange, would `remainingPatterns[0]` not suffice here? :)

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	for i := index; i < patternLen; i++ {
+		if pattern[i] == '\\' {
+			i += 1
+			if i == patternLen {
```
I think this case is already covered in the function above, perhaps it would be cleaner just as a check for `strings.HasSuffix(pattern, '\')`?

olivercalder replied:
It's only covered above if the trailing `'\\'` does not occur in a group. But you're right, a single check at the beginning of `ExpandPathPattern` would eliminate the need for both these checks. I'll do that!

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+			continue
+		}
+		infix := pattern[currLiteralStart:i]
+		groupExpanded, groupEnd, err := expandPathPatternFromIndex(pattern, i+1, patternLen)
```
Unless I'm mistaken, this looks like it'll expand for every non-special character? As in `/abc` will call it multiple times despite no groups being present?

olivercalder replied:
The check above is `if pattern[i] != '{'`. I didn't want the complex path to be in an if statement, so there's an early `continue` if it's a normal non-group case.

TheSignPainter98 replied:
Ah right my mistake

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	newExpanded := make([]string, 0, len(expanded))
+	suffix := pattern[currLiteralStart:patternLen]
+	for _, prefix := range expanded {
+		final := cleanPattern(prefix + suffix)
```
I'd avoid final in case this needs to be changed later or processed by the caller. Perhaps `cleaned` would work?

olivercalder replied:
Good point, thanks!

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	// Append trailing literal string, if any, to all previously-expanded
+	// patterns, and clean the resulting patterns.
+	alreadySeen := make(map[string]bool, len(expanded))
+	newExpanded := make([]string, 0, len(expanded))
```
I think it may be possible to modify `expanded` in place here

olivercalder replied:
It is possible, but then there would be lots of re-allocations if repeated appending exceeds the capacity multiple times. I'm essentially trying to grow the capacity of `expanded`, and afaik, the best way to do so is to allocate a new slice, `copy()`, then reassign the original variable to the new slice.

I'll move `expanded = newExpanded` to right after the `copy()`, and then append to `expanded`.

TheSignPainter98 replied:
Ah there's a secret pattern here, if you use 
```go
expanded = append(expanded, make([]string, len(currPrefixes))
```
Go will recognise this and convert it into a single call to `runtime.GrowSlice(len(currPrefixes))`

TheSignPainter98 commented on overlord/ifacestate/apparmorprompting/common/common.go:
```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	return pattern
+}
+
+func shiftStarAfterAnyChars(s string) string {
```
Unless these are required elsewhere, I think using closures (inlined or not) could scope these more clearly

olivercalder replied:
Very good point, thanks!

TheSignPainter98 reviewed (commented):
Here's a first go, mostly nitpicks

olivercalder committed:
o/i/a/common: add validation functions for expiration

Adds expiration-related analogues for `ValidateLifespanParseDuration`
and `ValidateConstraintsOutcomeLifespanDuration`.

Additionally, improves error messages by including the contents of the
invalid field in the message.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/requestrules: store expanded path patterns for rules

When adding a rule to the rule storage, expand the path pattern into a
list of patterns which do not contain groups. Then, try to add each to
the relevant path map, and if there is an identical expanded path
already, then there is a conflicting rule.

Additionally, `RefreshTreeEnforceConsistency` now performs validation on
the constraints, outcome, lifespan, and expiration of the rule, and if
it is invalid, discards it. If a rule with lifespan timespan has an
expiration time in the past, discard it as well. Then, the path pattern
for the rule is expanded and attached to the rule contents as a
non-public non-json field. `RefreshTreeEnforceConsistency` should only
be called when loading rules from disk (where validation ought to
occur), so this does not impose an undue performance penalty during
usual operation.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/a/common: refactored after code review (thanks @TheSignPainter98)

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder commented:
I'm going to close this now in favor of a full PR to merge the `common` package into snapd master. I'll make a note there that there are some unaddressed comments left over here, and I'd appreciate any future comments be posted there instead. Thanks!
